File: YandexTraining/AlgorithmsTraining/Cpp-AlgorithmsTraining/Warmup/G.cpp
#include <iostream>
#include <vector>

typedef std::vector<std::vector<int>> vvint;

int minorSum(int leftUpperCornerI, int leftUpperCornerJ, int sideLength, vvint field) {
    int sum = 0;
    for (int i = leftUpperCornerI; i - leftUpperCornerI < sideLength; i++) {
        for (int j = leftUpperCornerJ; j - leftUpperCornerJ < sideLength; j++) {
            sum += field[i][j];
        }
    }
    return sum;
}

int main() {
    int rows, columns, rank = 0;
    std::cin >> rows >> columns;
    vvint field;
    for (int i = 0; i < rows; i++) {
        field.push_back(std::vector<int>(columns));
        for (int j = 0; j < columns; j++) std::cin >> field[i][j];
    }

    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < columns; j++) {
            if (!field[i][j]) continue;
            if (!rank) rank++;
            int sum = 1;

            for (int k = 2; k + i <= rows && k + j <= columns; k++) {
                sum = minorSum(i, j, k, field);
                if (sum != k * k) break;
                if (rank < k) rank = k;
            }
        }
    }
    std::cout << rank;

    return EXIT_SUCCESS;
}

File: YandexTraining/AlgorithmsTraining/Python-AlgorithmsTraining/Warmup/G.py
def minorSum(leftUpperCornerI, leftUpperCornerJ, sideLength, field):
    sum = 0
    i = leftUpperCornerI
    while i - leftUpperCornerI < sideLength:
        j = leftUpperCornerJ
        while j - leftUpperCornerJ < sideLength:
            sum += field[i][j]
            j += 1
        i += 1
    return sum


def main():
    rows, columns = map(int, input().split())
    field = [list(map(int, input().split())) for i in range(rows)]
    rank = 0

    for i in range(rows):
        for j in range(columns):
            if not field[i][j]:
                continue
            if not rank:
                rank += 1

            sum = 1
            k = 2

            while k + i <= rows and k + j <= columns:
                sum = minorSum(i, j, k, field)
                if sum != k * k:
                    break
                if rank < k:
                    rank = k
                k += 1

    print(rank)


if __name__ == '__main__':
    main()
